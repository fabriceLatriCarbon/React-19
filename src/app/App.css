#root {
  max-width: 1280px;
  width: calc(100vw - 20%);
  margin: 0 auto;
  padding: 2rem;
  text-align: center;
}
