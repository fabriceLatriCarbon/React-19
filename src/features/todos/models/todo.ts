

export type Todo = {
  id: string,
  order: number,
  title: string,
  completed: boolean,
}