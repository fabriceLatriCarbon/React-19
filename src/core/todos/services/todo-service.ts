import { SERVER_TIMEOUT_DELAY } from "../../http-client/constants";
import { getHttpClient } from "../../http-client";
import { IHttpClient } from "../../http-client/contracts";
import { Todo } from "../models/todo";
import { ApiRoutes } from "./APIs/apiRoutes";
import { AddNewTodoRequest } from "src/features/todos/addTodo/models/DTOs/requests/add-new-todo";
import { AddNewTodoResponse } from "src/features/todos/addTodo/models/DTOs/responses/add-new-todo";
import { DeleteTodoResponse } from "src/features/todos/getTodos/models/DTOs/responses/delete-todo";
import { GetAllTodosResponse } from "src/features/todos/deleteTodo/models/DTOs/responses/get-all-todos";

export async function getAllTodos(): Promise<Todo[]> {
  const httpClient: IHttpClient = getHttpClient();

  if (!httpClient) throw new Error('httpClient must be provided');

  const todos = (await httpClient.get<GetAllTodosResponse>(ApiRoutes.GET_TODOS)).todos;

  return todos;
}

export async function addNewTodo(title: string): Promise<Todo> {
  'use server';
  
  return new Promise((resolve) => {
    setTimeout(async () => {
      const httpClient: IHttpClient = getHttpClient();
    
      if (!httpClient) throw new Error('httpClient must be provided');
    
      const newTodo = (await httpClient.post<AddNewTodoResponse, AddNewTodoRequest>(ApiRoutes.ADD_NEW_TODO, { title })).todo;
    
      resolve(newTodo);
    }, SERVER_TIMEOUT_DELAY);
  })
}

export async function deleteTodo(id: string): Promise<string> {
  'use server';
  const httpClient: IHttpClient = getHttpClient();

  if (!httpClient) throw new Error('httpClient must be provided');

  return (await httpClient.delete<DeleteTodoResponse>(ApiRoutes.DELETE_TODO, { todoId: id })).id
}

export function getTodoHighOrder(todos: Todo[]): number {
  return todos.toSorted((ta, tb) => ta.order < tb.order ? 1 : -1).at(0)?.order ?? 0;
}


export function createTodoFactory(order: number, title: string): Todo {
  return { id:Date.now().toString(), order: order + 1, title, completed: false }
}