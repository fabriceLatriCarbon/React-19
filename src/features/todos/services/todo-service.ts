import { SERVER_TIMEOUT_DELAY } from "../../../core/http-client/constants";
import { getHttpClient } from "../../../core/http-client";
import { IHttpClient } from "../../../core/http-client/contracts";
import { AddNewTodoRequest } from "../models/DTOs/requests/add-new-todo";
import { AddNewTodoResponse } from "../models/DTOs/responses/add-new-todo";
import { DeleteTodoResponse } from "../models/DTOs/responses/delete-todo";
import { GetAllTodosResponse } from "../models/DTOs/responses/get-all-todos";
import { Todo } from "../models/todo";
import { ApiRoutes } from "./APIs/routes";

export async function getAllTodos(): Promise<Todo[]> {
  const httpClient: IHttpClient = getHttpClient();

  if (!httpClient) throw new Error('httpClient must be provided');

  const todos = (await httpClient.get<GetAllTodosResponse>(ApiRoutes.GET_TODOS)).todos;

  return todos;
}

export async function addNewTodo(title: string): Promise<Todo> {
  'use server';
  
  return new Promise((resolve) => {
    setTimeout(async () => {
      const httpClient: IHttpClient = getHttpClient();
    
      if (!httpClient) throw new Error('httpClient must be provided');
    
      const newTodo = (await httpClient.post<AddNewTodoResponse, AddNewTodoRequest>(ApiRoutes.ADD_NEW_TODO, { title })).todo;
    
      resolve(newTodo);
    }, SERVER_TIMEOUT_DELAY);
  })
}

export async function deleteTodo(id: string): Promise<string> {
  'use server';
  const httpClient: IHttpClient = getHttpClient();

  if (!httpClient) throw new Error('httpClient must be provided');

  return (await httpClient.delete<DeleteTodoResponse>(ApiRoutes.DELETE_TODO, { todoId: id })).id
}

export function getTodoHighOrder(todos: Todo[]): number {
  return todos.toSorted((ta, tb) => ta.order < tb.order ? 1 : -1).at(0)?.order ?? 0;
}


export function createTodoFactory(order: number, title: string): Todo {
  return { id:Date.now().toString(), order: order + 1, title, completed: false }
}