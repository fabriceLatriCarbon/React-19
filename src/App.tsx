import './App.css'
import TodoApp from './features/todos/components/TodoApp'
import TodoProvider from './features/todos/contexts/todoContext';
import { getAllTodos } from './features/todos/services/todo-service';


async function App() {
  const todos = await getAllTodos();

  return (
    <TodoProvider todos={todos} >
      <TodoApp />
    </TodoProvider>
  )
}

export default App
