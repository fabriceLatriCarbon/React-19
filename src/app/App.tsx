import TodoProvider from '../core/todos/contexts/todoContext';
import { getAllTodos } from '../core/todos/services/todo-service';
import TodoPage from '../pages/TodoPage';
import './App.css'


async function App() {
  const todos = await getAllTodos();

  return (
    <TodoProvider todos={todos} >
      <TodoPage />
    </TodoProvider>
  )
}

export default App
