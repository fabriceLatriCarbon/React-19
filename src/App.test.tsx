import { render } from '@testing-library/react';
import App from "./App"
import { getAllTodos } from './features/todos/services/todo-service';
import { Todo } from './features/todos/models/todo';

const mockTodos: Todo[] = [
  {
    id: Date.now().toString(),
    order: 1,
    title: "First Test Todo",
    completed: false
  },
  {
    id: (Date.now() + 1).toString(),
    order: 2,
    title: "Second Test Todo",
    completed: false
  }
]

jest.mock('./features/todos/services/todo-service');

const mockGetAllTodos = jest.mocked(getAllTodos);

describe('App', () => {

  it('should render the app', async () => {
    mockGetAllTodos.mockResolvedValue(mockTodos);
    
    const { container } = render(await App());

    expect(container).toMatchSnapshot();
  })
})

