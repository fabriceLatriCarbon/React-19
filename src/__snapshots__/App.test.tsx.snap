// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render the app 1`] = `
<div>
  <form
    action="javascript:throw new Error('A React form was unexpectedly submitted. If you called form.submit() manually, consider using form.requestSubmit() instead. If you\\'re trying to use event.stopPropagation() in a submit event handler, consider also calling event.preventDefault().')"
    class="todo-form-container"
  >
    <div
      class="todo-form-main"
    >
      <div
        class="todo-form-input-container"
      >
        <label
          class="todo-form-input-label"
          for="todo-input"
        >
          Write something...
        </label>
        <input
          class="todo-form-input-txt"
          id="todo-input"
          name="todo-input"
        />
      </div>
      <button
        class="todo-form-submit-btn"
        id="todo-form-submit-btn"
      >
        Add
      </button>
    </div>
  </form>
  <div
    class="todo-list-container"
  >
    <div
      class="todo-item-container"
    >
      <p
        class="todo-item-title"
      >
        First Test Todo
         
      </p>
      <button
        class="todo-item-delete-btn"
      >
        X
      </button>
    </div>
    <div
      class="todo-item-container"
    >
      <p
        class="todo-item-title"
      >
        Second Test Todo
         
      </p>
      <button
        class="todo-item-delete-btn"
      >
        X
      </button>
    </div>
  </div>
</div>
`;
